chore: convert luhn algorithm project to lab (#744)
freeCodeCamp/CurriculumExpansion: fullstack-cert/python-projects/luhn-algo-lab/user-stories.md

- Starting from right, and excluding the rightmost digit (the check digit), double the value of every other digit.
- If the result of doubling a digit is greater than `9`, sum the digits to get a single digit. Alternatively, you can subtract `9` from the result.
- Take the sum of all the digits including the check digit.
- If the sum of all the digits is a multiple of `10`, then the number is valid; else it is not valid.

Let's say we have the number `453914880`. The steps to validate it using the Luhn algorithm would be:

```md
Account number       4   5   3   9   1   4   8   8   0   
Double every other       ✔       ✔       ✔       ✔   x 
                     4   10  3   18  1   8   8   16  0  
Sum 2-char digits    4    1  3    9  1   8   8    7  0
Total Sum:           4 + 1 + 3 + 9 + 1 + 8 + 8 + 7 + 0 = 41
```

- Since `41` is not a multiple of `10`, the number is invalid.

In this lab, you will build a credit card validator using the Luhn algorithm.

## User Stories

> 1. You should define a function named `verify_card_number` that takes a string of digits (representing a card number) as input and returns `True` if the card number is valid according to the Luhn algorithm, and `False` otherwise.

Reviewer: Looking at the code, this is no longer true. Do we want to return a boolean, print `VALID!`/`INVALID!`, or both?

Reviewer 2: I'd suggest returning `true`, `false` instead of the strings.

1. Within the `verify_card_number` function:

>    - You should remove any dashes or spaces from the card number before passing it to `verify_card_number`.

Reviewer: Actually, this has already been passed to `verify_card_number`
```suggestion
   - You should handle any dashes or spaces that may be present in the card number passed to `verify_card_number`.
```

The Luhn algorithm, also known as the "modulus 10" or "mod 10" algorithm, is a simple checksum formula used to validate a variety of identification numbers, like credit card numbers. These are the steps to validate a number using the Luhn algorithm:

> - Set aside the rightmost digit (the check digit).
> - Excluding the last digit, start from the rightmost digit and double the value of every second digit.

Reviewer: ```suggestion
- Starting from right, and excluding the rightmost digit (the check digit), double the value of every other digit.
```
Repeating the part of excluding the rightmost digit twice can be confusing

- If the result of doubling a digit is greater than `9`, sum the digits to get a single digit. Alternatively, you can subtract `9` from the result.
- Take the sum of all the digits including the check digit.
- If the sum of all the digits is a multiple of `10`, then the number is valid; else it is not valid.

Let's say we have the number `453914880`. The steps to validate it using the Luhn algorithm would be:

```md
Account number       4   5   3   9   1   4   8   8   0   
Double every other       ✔       ✔       ✔       ✔   x 
                     4   10  3   18  1   8   8   16  0  
Sum 2-char digits    4    1  3    9  1   8   8    7  0
Total Sum:           4 + 1 + 3 + 9 + 1 + 8 + 8 + 7 + 0 = 41
```

- Since `41` is not a multiple of `10`, the number is invalid.

In this lab, you will build a credit card validator using the Luhn algorithm.

## User Stories

> 1. You should define a function named `verify_card_number` to implement the Luhn algorithm for card number validation.
> 
> 1. The `verify_card_number` function should take a string of digits (representing a card number) as input and return a boolean value indicating whether the card number is valid according to the Luhn algorithm.
> 
> 1. To implement the Luhn algorithm within the `verify_card_number` function, you should:
>    - Start doubling every second digit from the right (excluding the check digit).
>    - If the result is `10` or more, sum the individual digits of the result (or subtract `9`).
>    - Sum all the digits together including the check digit.
>    - Return `True` if the final result is divisible by `10`; otherwise, return `False`.

Reviewer: ```suggestion
1. You should define a function named `verify_card_number` that takes a string of digits (representing a card number) as input and returns `True` if the card number is valid according to the Luhn algorithm, and `False` otherwise.
```
The part related to how to implement the algorithm is already explained before.

1. You should define a `main` function to demonstrate the use of `verify_card_number`.

1. In the `main` function:

>    - You should remove any dashes or spaces from the card number before passing it to `verify_card_number`.

Reviewer: I'm in favor of handling the removal of dashes and spaces within `verify_card_number` itself and remove `main`. It wouldn't be  possible to test `main` because we cannot pass a card number to validate

>    - If the result is valid, print `"VALID!"`; otherwise, print `"INVALID!"`.

Reviewer: ```suggestion
   - If the result is valid, print `VALID!`; otherwise, print `INVALID!`.
```
